### PR TITLE
fix(redis): handle rowcount when BaseException is raised

### PR DESCRIPTION
**Pull request by sileht**
In case of Redis raised a BaseException like asyncio.CancelledError.

ddtrace integration tries to compute the rowcount while "result"
variable does not exists.

```
  File "ddtrace/contrib/redis_utils.py", line 72, in _run_redis_command_async
    result = await func(*args, **kwargs)
  File "redis/asyncio/client.py", line 605, in execute_command
    conn = self.connection or await pool.get_connection(command_name, **options)
  File "redis/asyncio/connection.py", line 1076, in get_connection
    await self.ensure_connection(connection)
  File "redis/asyncio/connection.py", line 1115, in ensure_connection
    if await connection.can_read_destructive():
  File "redis/asyncio/connection.py", line 504, in can_read_destructive
    return await self._parser.can_read_destructive()
  File "redis/_parsers/hiredis.py", line 179, in can_read_destructive
    return await self.read_from_socket()
  File "redis/_parsers/hiredis.py", line 184, in read_from_socket
    buffer = await self._stream.read(self._read_size)
  File "asyncio/streams.py", line 713, in read
    await self._wait_for_data('read')
  File "asyncio/streams.py", line 545, in _wait_for_data
    await self._waiter
CancelledError: null

  File "xxxxxxx.py", line 287, in list_pulls
    await redis.zrangebyscore(
  File "ddtrace/contrib/redis/asyncio_patch.py", line 20, in traced_async_execute_command
    return await _run_redis_command_async(span=span, func=func, args=args, kwargs=kwargs)
  File "ddtrace/contrib/redis_utils.py", line 79, in _run_redis_command_async
    rowcount = determine_row_count(redis_command=redis_command, result=result)
UnboundLocalError: cannot access local variable 'result' where it is not associated with a value
```

This change fixes that. 

Fixes https://github.com/DataDog/dd-trace-py/issues/9548

## Checklist

- [X] Change(s) are motivated and described in the PR description
- [x] Testing strategy is described if automated tests are not included in the PR
- [X] Risks are described (performance impact, potential for breakage, maintainability)
- [x] Change is maintainable (easy to change, telemetry, documentation)
- [x] [Library release note guidelines](https://ddtrace.readthedocs.io/en/stable/releasenotes.html) are followed or label `changelog/no-changelog` is set
- [x] Documentation is included (in-code, generated user docs, [public corp docs](https://github.com/DataDog/documentation/))
- [x] Backport labels are set (if [applicable](https://ddtrace.readthedocs.io/en/latest/contributing.html#backporting))
- [x] If this PR changes the public interface, I've notified `@DataDog/apm-tees`.

## Reviewer Checklist

- [x] Title is accurate
- [x] All changes are related to the pull request's stated goal
- [x] Description motivates each change
- [x] Avoids breaking [API](https://ddtrace.readthedocs.io/en/stable/versioning.html#interfaces) changes
- [x] Testing strategy adequately addresses listed risks
- [x] Change is maintainable (easy to change, telemetry, documentation)
- [x] Release note makes sense to a user of the library
- [x] Author has acknowledged and discussed the performance implications of this PR as reported in the benchmarks PR comment
- [x] Backport labels are set in a manner that is consistent with the [release branch maintenance policy](https://ddtrace.readthedocs.io/en/latest/contributing.html#backporting)

### CODE REVIEW

**Comment by brettlangdon** [comment]
Thanks @sileht, is there a regression test we can add to simulate this behavior?

**Comment by sileht** [comment]
> Thanks @sileht, is there a regression test we can add to simulate this behavior?

@brettlangdon I added one.

**Comment by majorgreys** [comment]
@sileht lgtm - just needs a release note for the fix:

https://ddtrace.readthedocs.io/en/stable/releasenotes.html

File: ddtrace/contrib/redis_utils.py [hunk 1]
```diff
@@ -71,7 +71,7 @@ async def _run_redis_command_async(ctx: core.ExecutionContext, func, args, kwarg
     try:
         result = await func(*args, **kwargs)
         return result
-    except Exception:
+    except BaseException:
```
Comment (emmettbutler) [inline ddtrace/contrib/redis_utils.py:74]: I'm not sure how this fixes the issue given that `BaseException` is a superclass of `UnboundLocalError`
```
>>> isinstance(UnboundLocalError(), BaseException)
True
```

Comment (sileht) [reply ddtrace/contrib/redis_utils.py:74]: It's `asyncio.CancelledError` that is raised within the `try/except`, not `UnboundLocalError`.
The fixes will catch it and the `UnboundLocalError` will not be raised inside the `except`.

The test proves it works.

**Comment by sileht** [comment]
@majorgreys @brettlangdon @P403n1x87 

any update on this?

**Review by emmettbutler — APPROVED** [review]

**Comment by aucampia** [comment]
Can someone re-trigger the CI? It seems like the error was transient. We would really like to have this fix, it is causing problems for us in production.

**Review by tabgok — APPROVED** [review]

**Pull request merged**